clients/da-vinci-client/src/main/java/com/linkedin/davinci/client/DaVinciRecordTransformer.java

```diff
@@ -96,10 +163,112 @@ public final ByteBuffer getValueBytes(Schema schema, V value) {
   }
 
   /**
-   *
-   * @return the storeVersion
+   * @return {@link #storeVersion}
    */
   public final int getStoreVersion() {
     return storeVersion;
   }
+
+  /**
+   * @return the hash of the class bytecode
+   */
+  // Visible for testing
+  public final int getClassHash() {
+    String className = this.getClass().getName().replace('.', '/') + ".class";
+    try (InputStream inputStream = this.getClass().getClassLoader().getResourceAsStream(className)) {
+      ClassReader classReader = new ClassReader(inputStream);
+      byte[] bytecode = classReader.b;
+      return Arrays.hashCode(bytecode);
+
+    } catch (IOException e) {
+      throw new VeniceException("Failed to get classHash", e);
+    }
+  }
+
+  /**
+   * @return true if the transformation logic has changed since the last time the class was loaded
+   */
+  // Visible for testing
+  public final boolean hasTransformationLogicChanged(int classHash) {
+    try {
+      String classHashPath = String.format("./classHash-%d.txt", storeVersion);
+      File f = new File(classHashPath);
+      if (f.exists()) {
+        try (BufferedReader br = new BufferedReader(new FileReader(classHashPath))) {
+          int storedClassHash = Integer.parseInt(br.readLine());
+          if (storedClassHash == classHash) {
+            return false;
+          }
+        }
+      }
+
+      try (FileWriter fw = new FileWriter(classHashPath)) {
+        fw.write(String.valueOf(classHash));
+      }
+      return true;
+    } catch (IOException e) {
+      throw new VeniceException("Failed to check if transformation logic has changed", e);
+    }
+  }
+
+  /**
+   * Bootstraps the client after it comes online and {@link #onStart()} completes.
+   */
+  public final void onRecovery(
+      AbstractStorageEngine storageEngine,
+      StoreBackend storeBackend,
+      BlobTransferManager blobTransferManager,
+      List<Integer> partitions,
+      Optional<Version> version) {
+
+    if (blobTransferManager != null) {
```

Comment:
Can you elaborate for me a bit on why this must be the case?  Looking at the below code, it either bootstraps from VT or local storage.  It would seem that if blob transfer is available, it would just create the local storage and then we would read that.

Reply:
Blob transfer would require some modifications to its API to get it working with the record transformer, such as passing in the classHash into the blob transfer's request body. However, since work on the blob transfer has been paused, I was advised by Manu to throw an error in the meantime before work is resumed on it.

Reply:
The hash is stored in rocksdb right?  In which case, why would it need to be part of the api?  It should just transfer

Reply:
Right now, the hash is being written directly to a directory on disk, but it would simplify things if it was just written to RocksDB.

Reply:
Ok.  Can you size that task?  I know we don't have a ton of official cycles we can commit here.  If it's a good amount of refactoring to move the hash to rocksdb then I'm content to just leave this as a TODO item.

Reply:
I think it should be a fairly small task. Is not just calling `storageEngine::put` and `storageEngine::get`?